src/openms/source/ANALYSIS/ID/PrecursorPurity.cpp

```diff
@@ -0,0 +1,231 @@
+    if (score.target_intensity > 0.0) // otherwise default value of 0 is used
+    {
+      score.signal_proportion = score.target_intensity / score.total_intensity;
+    }
+    score.target_peak_count = score1.target_peak_count + score2.target_peak_count;
+    score.residual_peak_count = score1.residual_peak_count + score2.residual_peak_count;
+
+    return score;
+  }
+
+  std::vector<PrecursorPurity::PurityScores> PrecursorPurity::computePrecursorPurities(const PeakMap& spectra, double precursor_mass_tolerance, bool precursor_mass_tolerance_unit_ppm)
+  {
+    std::vector<PrecursorPurity::PurityScores> purityscores;
+
+    // TODO throw an exception or at least a warning?
+    // if there is no MS1 before the first MS2, the spectra datastructure is not suitable for this function
+    if (spectra[0].getMSLevel() == 2)
+    {
+      LOG_WARN << "Warning: Input data not suitable for Precursor Purity computation. Will be skipped!\n";
+      return purityscores;
+    }
+
+    // keep the index of the two MS1 spectra flanking the current group of MS2 spectra
+    Size current_parent_index = 0;
+    Size next_parent_index = 0;
+    bool lastMS1(false);
+    for (Size i = 0; i < spectra.size(); ++i)
+    {
+      // change current parent index if a new MS1 spectrum is reached
+      if (spectra[i].getMSLevel() == 1)
+      {
+        current_parent_index = i;
+      }
+      else if (spectra[i].getMSLevel() == 2)
+      {
+        // update next MS1 index, if it is lower than the current MS2 index
+        if (next_parent_index < i)
+        {
+          for (Size j = i+1; j < spectra.size(); ++j)
+          {
+            if (spectra[j].getMSLevel() == 1)
+            {
+              next_parent_index = j;
+              break;
+            }
+          }
+          // if the next MS1 index was not updated,
+          // the end of the PeakMap was reached right after this current group of MS2 spectra
+          if (next_parent_index < i)
+          {
+            lastMS1 = true;
+          }
+        }
+
+        // std::cout << "MS1 Spectrum: " << spectra[current_parent_index].getNativeID() << " | MS2 : " << spectra[i].getNativeID() << std::endl;
+        PrecursorPurity::PurityScores score1 = PrecursorPurity::computePrecursorPurity(spectra[current_parent_index], spectra[i].getPrecursors()[0], precursor_mass_tolerance, precursor_mass_tolerance_unit_ppm);
+        // use default values of 0, if there is an MS2 spectrum without an MS1 after it (may happen for the last group of MS2 spectra)
+        PrecursorPurity::PurityScores score2;
+        if (!lastMS1) // there is an MS1 after this MS2
+        {
+          score2 = PrecursorPurity::computePrecursorPurity(spectra[next_parent_index], spectra[i].getPrecursors()[0], precursor_mass_tolerance, precursor_mass_tolerance_unit_ppm);
+        }
+        PrecursorPurity::PurityScores score = PrecursorPurity::combinePrecursorPurities(score1, score2);
+        purityscores.push_back(score);
```

Comment:
how do you link the scores to the MS spectra? It seems it would make sense to have an index / identifier to link the two?

Reply:
true, currently they are linked only by their order, the 5th `PurityScore` belongs to the 5th MS2 spectrum in the map (not counting the MS1 spectra).
That is slightly annoying to deal with. But to simplify it I would rather need some kind of index in the `PeakMap` that points to the correct PrecursorPurity. In my search engine I have some filtering steps, that remove the MS1 spectra and some useless MS2 spectra after running this and I have to keep track of which spectra were discarded. So keeping a spectrum index in `PurityScore` would be useless. The most convenient solution for me would be to add this whole data structure to the `Precursor` class, but I don't want to add stuff there, that only a few people will need.